streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverSchedulingPolicy.scala

```diff
@@ -0,0 +1,161 @@
+        hostToExecutors.get(host) match {
+          case Some(executorsOnHost) =>
+            // preferredLocation is a known host. Select an executor that has the least receivers in
+            // this host
+            val scheduledLocation =
+              executorsOnHost.minBy(executor => numReceiversOnExecutor(executor))
+            locations(i) += scheduledLocation
+            numReceiversOnExecutor(scheduledLocation) =
+              numReceiversOnExecutor(scheduledLocation) + 1
+          case None =>
+            // preferredLocation is an unknown host.
+            // Note: There are two cases:
+            // 1. This executor is not up. But it may be up later.
+            // 2. This executor is dead, or it's not a host in the cluster.
+            // Currently, simply add host to the scheduled locations
+            locations(i) += host
+        }
+      }
+    }
+
+    // For those receivers that don't have preferredLocation, make sure we assign at least one
+    // executor to them.
+    for (scheduledLocations <- locations.filter(_.isEmpty)) {
+      // Select the executor that has the least receivers
+      val (executor, numReceivers) = numReceiversOnExecutor.minBy(_._2)
+      scheduledLocations += executor
+      numReceiversOnExecutor(executor) = numReceivers + 1
+    }
+
+    // Assign idle executors to receivers that have less executors
+    val idleExecutors = numReceiversOnExecutor.filter(_._2 == 0).map(_._1)
+    for (executor <- idleExecutors) {
+      // Assign an idle executor to the receiver that has least locations.
+      val scheduledLocations = locations.minBy(_.size)
+      scheduledLocations += executor
+    }
+
+    receivers.map(_.streamId).zip(locations).toMap
+  }
+
+  /**
+   * Return a list of candidate executors to run the receiver. If the list is empty, the caller can
+   * run this receiver in arbitrary executor.
+   *
+   * This method tries to balance executors' load. Here is the approach to schedule executors
+   * for a receiver.
+   * <ol>
+   *   <li>
+   *     If preferredLocation is set, preferredLocation should be one of the candidate executors.
+   *   </li>
+   *   <li>
+   *     Every executor will be assigned to a weight according to the receivers running or
+   *     scheduling on it.
+   *     <ul>
+   *       <li>
+   *         If a receiver is running on an executor, it contributes 1.0 to the executor's weight.
+   *       </li>
+   *       <li>
+   *         If a receiver is scheduled to an executor but has not yet run, it contributes
+   *         `1.0 / #candidate_executors_of_this_receiver` to the executor's weight.</li>
+   *     </ul>
+   *     At last, if there are more than 3 idle executors (weight = 0), returns all idle executors.
+   *     Otherwise, we only return 3 best options according to the weights.
+   *   </li>
+   * </ol>
+   *
+   * This method is called when a receiver is registering with ReceiverTracker or is restarting.
+   */
+  def rescheduleReceiver(
```

Comment:
This method was called `scheduleReceiver`. I renamed it to `rescheduleReceiver` to avoid confusing with `scheduleReceivers`.